offsite-activity-signup/Code.gs

```diff
@@ -0,0 +1,346 @@
+  writeAttendeeAssignments_(ss, attendees);
+  writeActivityRosters_(ss, activities);
+}
+
+/**
+ * Select activities via random priority.
+ */
+function assignWithRandomPriority_(attendees, activities, numActivitiesPerPerson) {
+  var activitiesById = activities.reduce(function(obj, activity) {
+    obj[activity.id] = activity;
+    return obj;
+  }, {});
+  for (var i = 0; i < numActivitiesPerPerson; ++i) {
+    var randomizedAttendees = shuffleArray_(attendees);
+    randomizedAttendees.forEach(function(attendee) {
+      makeChoice_(attendee, activitiesById);
+    });
+  }
+}
+
+/**
+ * Attempt to assign an activity for an attendee based on their preferences
+ * and current schedule.
+ */
+function makeChoice_(attendee, activitiesById) {
+  for(var i in attendee.preferences) {
+    var activity = activitiesById[attendee.preferences[i]];
+    if (!activity) {
+      continue;
+    }
+    var canJoin = checkAvailability_(attendee, activity);
+    if (canJoin) {
+      attendee.assigned.push(activity);
+      activity.roster.push(attendee);
+      return;
+    }
+  }
+}
+
+/**
+ * Checks that an activity has capacity and doesn't conflict with
+ * previously assigned activities.
+ */
+function checkAvailability_(attendee, activity) {
+  if (activity.capacity <= activity.roster.length) {
+    return false;
+  }
+  var timesConflict = attendee.assigned.some(function(assignedActivity) {
+    return !(assignedActivity.startAt.getTime() > activity.endAt.getTime() || activity.startAt.getTime() > assignedActivity.endAt.getTime());
+  });
+  if (timesConflict) {
+    return false;
+  }
+  return true;
+};
+
+/**
+ * Populates a sheet with the assigned activities for each attendee.
+ */
+function writeAttendeeAssignments_(ss, attendees) {
+  var sheetName = 'Activities by person';
+  var sheet = ss.getSheetByName(sheetName);
+  if (!sheet) {
+    sheet = ss.insertSheet(sheetName);
+  }
+  sheet.clear();
+  sheet.appendRow(["Email address", "Activities"]);
+  sheet.getRange("B1:1").merge();
+  attendees.forEach(function(attendee) {
+    var assignedActivities = attendee.assigned.map(function(activity) {
+      return activity.description;
+    });
+    var row = [attendee.email].concat(assignedActivities);
+    sheet.appendRow(row);
+  });
+  sheet.setFrozenRows(1);
+  sheet.getRange("1:1").setFontWeight("bold");
+  sheet.autoResizeColumns(1, sheet.getLastColumn());
+}
+
+/**
+ * Populates a sheet with the rosters for each activity.
+ */
+function writeActivityRosters_(ss, activities) {
+  var sheetName = 'Activity rosters';
+  var sheet = ss.getSheetByName(sheetName);
+  if (!sheet) {
+    sheet = ss.insertSheet(sheetName);
+  }
+  sheet.clear();
+  var rows = [];
+  // Format each roster as a column in the sheet
+  activities.forEach(function(activity, colIndex) {
+    if (!rows[0]) {
+      rows[0] = fillArray_([], activities.length, '');
+    }
+    rows[0][colIndex] = activity.description;
+    activity.roster.forEach(function(attendee, rowIndex) {
```

Comment:
Optional: Groking this code was more difficult than ideal. Would it be cleaner to:

1. Assemble a 2D array of attendees, where each row is one activity:
```
[
  // Activity 1
  [person 1, person3]
  // Activity 2
  [person 2, person 4]
]
```
2. Rotate the grid (using a helper function), so that rows become columns.
3. Add spaces to the grid to ensure it's square (using a helper function).

Reply:
Ha, that's what I had originally but thought that might be harder (those not that anyone needs to understand the transpose code :)

Switched back to that version.